fields/sql.go

```diff
@@ -104,3 +107,157 @@ func (f Valuer) Value() (driver.Value, error) {
 
 var _ driver.Valuer = Valuer{}
 var _ driver.Valuer = &Valuer{}
+
+// Scanner fulfills the sql.Scanner interface which deserializes SQL values
+// into the type dictated by our descriptor.
+type Scanner struct {
+	*Descriptor
+	value   reflect.Value
+	IsValid bool
+}
+
+// Interface returns the value deserialized into the scanner.
+func (s Scanner) Interface() interface{} {
+	if s.value.IsValid() {
+		return s.value.Interface()
+	}
+	return nil
+}
+
+// Scan satisfies the sql.Scanner interface.
+// The src value will be one of the following:
+//    int64
+//    float64
+//    bool
+//    []byte
+//    string
+//    time.Time
+//    nil - for NULL values
+func (s *Scanner) Scan(src interface{}) error {
+	// Keep track of whether our value was empty.
+	s.IsValid = src != nil
+	// Get a value of the pointer of the pointer of our type. The Scanner and Unmarshalers should
+	// only be implemented as dereference methods, since they would do nothing otherwise. Therefore
+	// we can safely assume that we should check for these interfaces on the pointer value.
+	s.value = reflect.New(s.Type)
+	i := s.value.Interface()
+	// Our value however should continue referencing a non-pointer for easier assignment.
+	s.value = s.value.Elem()
+
+	// If our interface supports sql.Scanner we can immediately short-circuit as this is what the
+	// MySQL driver would do.
+	if scanner, ok := i.(sql.Scanner); ok {
+		return scanner.Scan(src)
+	}
+
+	// Null values are simply set to zero. Because we're not holding on to pointers, we need to
+	// represent this as a boolean. This comes _after_ the scanner step, just in case the scanner
+	// handles nil differently.
+	if !s.IsValid {
+		return nil
+	}
+
+	// Handle coercion into native types []byte and time.Time.
+	// These are pointers here because we want to pass around a pointer for interfaces
+	switch i.(type) {
+	case *[]byte:
+		if str, ok := src.(string); ok {
+			s.value.Set(reflect.ValueOf([]byte(str)))
+			return nil
+		}
+		if _, ok := src.([]byte); ok {
+			s.value.Set(reflect.ValueOf(src))
+			return nil
+		}
+	case *time.Time:
+		if _, ok := src.(time.Time); ok {
+			s.value.Set(reflect.ValueOf(src))
+			return nil
+		}
+	}
```

Comment:
could you comment on why these need to be handled in this codepath, but not in the valuer path?

Reply:
Thanks, I think that helps.